core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2496,8 +2496,23 @@ object Stream {
      * For example, `merge` is implemented by calling `unconsAsync` on each stream, racing the
      * resultant `AsyncPull`s, emitting winner of the race, and then repeating.
      */
-    def unconsAsync(implicit ec: ExecutionContext, F: Effect[F]): Pull[F,Nothing,AsyncPull[F,Option[(Segment[O,Unit], Stream[F,O])]]] =
-      Pull.fromFreeC(Algebra.unconsAsync(self.get, ec)).map(_.map(_.map { case (hd, tl) => (hd, Stream.fromFreeC(tl)) }))
+    def unconsAsync(implicit ec: ExecutionContext, F: Effect[F]): Pull[F,Nothing, AsyncPull[F,Option[(Segment[O,Unit], Stream[F,O])]]] = {
+      type UO = Option[(Segment[O,Unit], FreeC[Algebra[F,O,?],Unit])]
+
+      Pull.fromFreeC {
+        val ref = new async.Ref[F, Either[Throwable, Option[(Segment[O,Unit], Stream[F,O])]]]
```

Comment:
I would think this needs to be suspended in `F`, maybe it's ok since it's encapsulated?

Reply:
yes, the reason is to optimize as much as we can, hence this is evaluated at every async pull.

Reply:
:+1: